deepspeed/runtime/zero/stage3.py

```diff
@@ -1109,6 +1097,47 @@ def __reduce_and_partition_ipg_grads(self, safe_mode: bool = False) -> None:
             event.record()
             self.param_reduce_events.append(event)
 
+    @instrument_w_nvtx
+    def __avg_scatter_contiguous_grads(self, buffer_to_reduce: Tensor) -> List[Tensor]:
```

Comment:
What is the advantage of this over `__avg_scatter_grads()`?

Reply:
This version will call a single allreduce call on the flat grad buffer and then take the split the result into grad partition for each rank. It uses the knowledge the the gradients are stored in a flat buffer to skip building a flat buffer for reduction.
Compared to the standard version which builds a flat buffer from grad partitions by calling many narrow and copy operations and calling reduce_scatter.

Reply:
Thanks for the clarification. That makes sense. Could you please share some observed perf improvements?